### PR TITLE
Update mosaic_fsdp_utils.py

### PR DESCRIPTION
**Pull request by vchiley**
Porting https://github.com/mosaicml/llm-foundry/pull/1104 to composer.

# What does this PR do?

Updates how a new process group is initialized. This version was broken in some version of torch1.XX, but has been working well in torch2+

# What issue(s) does this change relate to?

- https://databricks.atlassian.net/browse/GRT-2813
- Porting https://github.com/mosaicml/llm-foundry/pull/1104 to composer.

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

**Comment by vchiley** [comment]
> LGTM! Assuming you tested?

Not yet. 
Theres a reason this is a draft 😅

**Comment by vchiley** [comment]
(I set really short warmups which explains some of the "loss spikes" that come back down)

Screenshot 2024-04-10 at 11 27 03 AM

The newest version is highlighted in aqua
Pink is the true baseline
Light blue is the baseline before https://github.com/mosaicml/llm-foundry/pull/1104

tldr lgtm.
